compilers/concrete-compiler/compiler/include/concretelang/TestLib/TestCircuit.h

```diff
@@ -0,0 +1,186 @@
+    return TestCircuit(clientCircuit, serverCircuit, useSimulation,
+                       artifactFolder, keyset);
+  }
+
+  TestCircuit(ClientCircuit clientCircuit, ServerCircuit serverCircuit,
+              bool useSimulation, Keyset keyset)
+      : clientCircuit(clientCircuit), serverCircuit(serverCircuit),
+        useSimulation(useSimulation), keyset(keyset) {}
+
+  Result<std::vector<Value>> call(std::vector<Value> inputs) {
+    auto preparedArgs = std::vector<TransportValue>();
+    for (size_t i = 0; i < inputs.size(); i++) {
+      OUTCOME_TRY(auto preparedInput, clientCircuit.prepareInput(inputs[i], i));
+      preparedArgs.push_back(preparedInput);
+    }
+    std::vector<TransportValue> returns;
+    if (useSimulation) {
+      OUTCOME_TRY(returns, serverCircuit.simulate(preparedArgs));
+    } else {
+      OUTCOME_TRY(returns, serverCircuit.call(keyset.server, preparedArgs));
+    }
+    std::vector<Value> processedOutputs(returns.size());
+    for (size_t i = 0; i < processedOutputs.size(); i++) {
+      OUTCOME_TRY(processedOutputs[i],
+                  clientCircuit.processOutput(returns[i], i));
+    }
+    return processedOutputs;
+  }
+
+  std::string getArtifactFolder() { return artifactFolder; }
+
+private:
+  TestCircuit(ClientCircuit clientCircuit, ServerCircuit serverCircuit,
+              bool useSimulation, std::string artifactFolder, Keyset keyset)
+      : clientCircuit(clientCircuit), serverCircuit(serverCircuit),
+        useSimulation(useSimulation), artifactFolder(artifactFolder),
+        keyset(keyset){};
+  ClientCircuit clientCircuit;
+  ServerCircuit serverCircuit;
+  bool useSimulation;
+  std::string artifactFolder;
+  Keyset keyset;
+};
+
+TestCircuit load(mlir::concretelang::CompilerEngine::Library compiled) {
+  auto keyset =
+      getTestKeySetCachePtr()
+          ->getKeyset(compiled.getProgramInfo().asReader().getKeyset(), 0, 0)
+          .value();
+  return TestCircuit::create(
+             keyset, compiled.getProgramInfo().asReader(),
+             compiled.getSharedLibraryPath(compiled.getOutputDirPath()), 0, 0,
+             false)
+      .value();
+}
+
+const std::string FUNCNAME = "main";
+
+std::string getSystemTempFolderPath() {
+  llvm::SmallString<0> tempPath;
+  llvm::sys::path::system_temp_directory(true, tempPath);
+  return std::string(tempPath);
+}
+
+std::string createTempFolderIn(const std::string &rootFolder) {
+  std::srand(std::time(nullptr));
+  auto new_path = [=]() {
+    llvm::SmallString<0> outputPath;
+    llvm::sys::path::append(outputPath, rootFolder);
+    std::string uid = std::to_string(
+        std::hash<std::thread::id>()(std::this_thread::get_id()));
+    uid.append("-");
+    uid.append(std::to_string(std::rand()));
+    llvm::sys::path::append(outputPath, uid);
+    return std::string(outputPath);
+  };
+
+  // Macos sometimes fail to create new directories. We have to retry a few
+  // times.
```

Comment:
This is weird. What's the error message when the creation fails?

Reply:
`Unknown Error` :)